### PR TITLE
Fix visual and DOM order of the list view header tabs and close button.

### LINKED ISSUE #58940: List View: tabs and close button visual order and DOM order mismatch
### Description

In the process of reviewing all the CSS `order` properties in use in the editor, I noticed the List View tabs and close button visual order and DOM order mismatch. For accessibility, the order must always match when it affects 'meaning and operation'. basically, only purely decorative elements can be reordered. Reference:

WCAG 2.2
1.3.2 Meaningful Sequence (Level A) https://www.w3.org/TR/WCAG22/#meaningful-sequence
2.4.3 Focus Order (Level A) https://www.w3.org/TR/WCAG22/#focus-order

Screenshot to illustrate the reading / DOM order mismatch.

Screenshot 2024-02-12 at 14 50 40

In this specific case, these elements are also interactive, which makes the problem even more relevant.

Instead of using `order`, the DOM order must be adjusted to match the desired visual order.

Introduced in https://github.com/WordPress/gutenberg/commit/a3a55bb21718833db88d516bb22edfd6021263ee so this `order` property is new on trunk and should be fixed for the WordPress 6.5 release.
Worth noting that the order mismatch occurs also in previous versione but for a different reason. The Close button used to be absollutely positioned.

The fix appears to be trivial as the button just needs to be moved after the tabs. Code:

https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/list-view-sidebar/index.js/#L133-L155

Since it appeaers to be a simple, self-contained fix, I'd like to propose it for WP 6.5 consideration.

### Step-by-step reproduction instructions

- Go to te post editor.
- Open the List VIew.
- Inspect the elements in the list view top part.
- Observe the Close button is first in the DOM but it's visually shown after the tabs.

Test with a screen reader, for example Safari + VoiceOver.
- Open the list view. Initial focus is set on the first item in the list (which is an a11y problem I will report separately).
- Press Control + Option + Left arrow key until you reach toe top part of the list view.
- Observe the first element VoiceOver moves to within the top part of the list view is the 'Outline' tab.
- Press Control + Option + Left arrow key again.
- Observe VoiceOver moves to the 'List View' tab.
- Press Control + Option + Left arrow key again.
- Observe VoiceOver moves to the Close button, unexpectedly jumping to the right.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/58940

## What?

In the List View header, the tabs and the close button visual order and DOM order mismatch. A new `order` CSS property was introduced to swap the order in place of the previous absolute positioning so, in a way, the problem is not new. However, it would be great to not introduce new `order' properties in the codebase.

## Why?

For accessibility, visual / reading order and DOM order must match when they affect 'meaning and operation'.

## How?

Render the elements in the desired order in the DOM rather than swapping their visual order via CSS.

## Testing Instructions

- Repeat the steps from the issue https://github.com/WordPress/gutenberg/issues/58940
- Observe the DOM order and the visual / tabbing order match.

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Thanks for the PR and the context @afercia. This is testing nicely for me, and doesn't appear to have any effect on the list view other than altering the tab sequence. I see that @andrewhayward added feedback on the approach in the linked issue here: https://github.com/WordPress/gutenberg/issues/58940#issuecomment-1939144357. Personally, I don't have a preference for which tab sequence in this particular case, I can see there are good arguments either way. One thing that will need to be updated prior to landing is that the e2e tests currently expect the behaviour that exists on trunk. E.g.

https://github.com/WordPress/gutenberg/blob/84715c01148344c318344978508b5d442526ed0c/test/e2e/specs/site-editor/list-view.spec.js#L116-L128

And: https://github.com/WordPress/gutenberg/blob/a2d973869205d71aa384e88a79032fe3cee5b66e/test/e2e/specs/editor/various/list-view.spec.js#L358-L370

I'll just add another couple of folks familiar with the list view accessibility in case they have opinions, too.

**Comment by talldan** [comment]
I wasn't particularly involved in it, but I'm sure I recall that the implementation in trunk was for accessibility reasons

It looks like @alexstine reworked the tabs (in [Add TabPanel to document overview replacing fake tabs](https://github.com/WordPress/gutenberg/pull/50199)), and in that PR the close button was positioned absolutely to appear on the right, but would be focused first via its position in the DOM order.

It was modified to use `order` in [Components: replace TabPanel with Tabs in the editor Document Overview sidebar](https://github.com/WordPress/gutenberg/pull/57082) by @chad1008.

Some feedback from those folks would be good.

**Comment by afercia** [comment]
@talldan thanks, I did check the PR where it was modified.
Ideally it would be best to have the Close button first in the DOM. However, if the current design can't be changed, we need to make the visual order match the DOM order. A mismatch is a no-go and it's a violation of two WCAG success criteria.
The feedback from @alexstins is valid from a perspective of a person with no vision but Alex is likely unaware of the visuals of the Link View header.

**Comment by talldan** [comment]
> The feedback from @alexstins is valid from a perspective of a person with no vision but Alex is likely unaware of the visuals of the Link View header.

That seems like an assumption. I personally think it'd be better to get feedback directly before undoing a change another contributor made.

**Review by andrewhayward — COMMENTED** [review]
I left [some feedback in the tracking issue](https://github.com/WordPress/gutenberg/issues/58940#issuecomment-1939144357), as it's more conceptual than specific to this fix.

**Review by alexstine — COMMENTED** [review]
@afercia Thanks for the PR. I agree this is something that should be fixed but not like this.

Is there any way to fix the visual order so you have close, tabs, then tab panel content? I do not want the tab order to find the close button before the tabs. This makes the tabs undiscoverable to keyboard users as they will be able to assume that there is no content before the close button as that is generally the first item in a region/dialog.

@talldan 

> That seems like an assumption. I personally think it'd be better to get feedback directly before undoing a change another contributor made.

Andrea is actually correct about this, I had no idea the visual order does not match and that is an accessibility WCAG failure. However, as I noted above, another solution needs to be found.

Thanks all.

**Comment by afercia** [comment]
@alexstine thanks for your feedback

>  I do not want the tab order to find the close button before the tabs.

Actually, this PR moves the Tabs before the Close button. The order is now:
- Tabs (List View and Outline).
- Close button.
- The active tab panel.

This matches the visual order.

However, in a Tabs interface there shouldn't be any focusable element between tabs and tabpanels. So yes, this PR fixes the order mismatch but it introduces a new problem.
Worth noting this is no different from the Settings panel.
The Settings panel has been recently refactored to use the Tabs component and now the order is:
- Tabs.
- Close button.
- The active tab panel.

That's not correct because there shouldn't be any focusable element between tabs and tab panels. That defeats the whole purpose of the Tabs interface which is allowing keyboard users to directly tab from the selected Tab to the active Tab panel. I will create a separate issue.

**Comment by afercia** [comment]
Any idea about how to solve the lates error?
```
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409)
Conflict: an artifact with this name already exists on the workflow run
```

**Comment by Mamaduka** [comment]
> Any idea about how to solve the lates error?

@afercia, the error is coming from the CI setup, unrelated to the e2e tests itself. Ref: https://github.com/WordPress/gutenberg/issues/58995.

**Comment by afercia** [comment]
Thanks @Mamaduka 👋🏽

**Comment by alexstine** [comment]
@afercia Yes, I got it backwards. Should be no close button in between tabs and tab panel. I wrote that too late I guess. Let's follow-up since this effects other areas of the editor.

**Comment by afercia** [comment]
I think the broader issue of not allowing extraneous tab stops between the Tabs and their panels should be addressed in https://github.com/WordPress/gutenberg/issues/59013 as it needs feedback from the coomponents package maintainers and from design.

In the meantime, I think fixing the order mismatch is worthwhile. I'd appreciate a final review, thanks.

**Comment by afercia** [comment]
I'd appreciate a final review and approval so this cna be cherry picked for 6.5 in https://github.com/WordPress/gutenberg/pull/59541

The broader issue is tracked at https://github.com/WordPress/gutenberg/issues/59013 and that should be addressed at the component level, IMHO.

**Review by alexstine — APPROVED** [review]
@afercia I can't really review CSS changes but assuming this is still an improvement. Let's follow-up on the other issue after this is merged. Thanks.

**Comment by andrewhayward** [comment]
I'm not going to block this, but wanted to voice my personal concern that I don't think this is an improvement, and if it were up to me I wouldn't be racing to get it in to 6.5.

[As noted in the supporting issue](https://github.com/WordPress/gutenberg/issues/58940#issuecomment-1939144357), I understand the motivation behind this PR, but I think the programmatically determined order actually makes more sense as things stand. Not only does the close button belong to the entire sidebar, rather than anything relating to the tabs, but now we're putting controls between those tabs and their associated content.

This is one of those murky "lesser of two evils" scenarios, and I'm not convinced we've picked the right one here.

**Comment by alexstine** [comment]
I agree with you @andrewhayward. Our other option is to leave a known accessibility WCAG failure. I really hate both and I tend to agree that the current trunk is a better experience for keyboard users. Any further thoughts? Maybe we forget this for now and make the proper solution for future release?

**Comment by youknowriad** [comment]
Given that RC1 is today, I'm removing this from 6.5.

**Comment by afercia** [comment]
> This is one of those murky "lesser of two evils" scenarios

I'd agree on this and I'm not opposed to close this PR as long as:
- https://github.com/WordPress/gutenberg/issues/59013 gets prioritized and a design change is accepted with no objections. The design team need to understand the current design is wrong. I'd greatly appreciate to not have to get into arguments about that. @andrewhayward I'd appreciate your full support on https://github.com/WordPress/gutenberg/issues/59013 thanks.
- The Tabs component is changed in a way that no one is ever able to add extraneous content between Tabs and Tabpanels. This requires changing the component implementation and I'd greatly appreciate to not have to get into philosophical arguments about it. The current Tabs component is basically open to misuse and that needs to be fixed. I'd greatly appreciate full support on this as well.

**Comment by andrewhayward** [comment]
> ...The design team need to understand the current design is wrong. I'd greatly appreciate to not have to get into arguments about that...

I'm quite happy to support the conversation, and to push for improvements (there shouldn't be focus breaks between tabs and panels, for example), but lets be mindful of how we discuss the project and the work that goes into it. It may be a small thing, but the project should be able to evolve safely; the designs aren't "wrong", they just need work. Not to start an argument (!), but a lot of the accessibility specs are equally subjective!

> * The Tabs component is changed in a way that no one is ever able to add extraneous content between Tabs and Tabpanels. This requires changing the component implementation and I'd greatly appreciate to not have to get into philosophical arguments about it. The current Tabs component is basically open to misuse and that needs to be fixed. I'd greatly appreciate full support on this as well.

I'm not sure as this will ever be possible, based purely on how the admin interface is built. There will almost certainly always be times where the tabs and their associated panels have to be inserted into different parts of the DOM tree.

**Pull request closed without merging**
